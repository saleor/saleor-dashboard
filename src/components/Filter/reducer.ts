import { update } from "@saleor/utils/lists";
import { IFilter, IFilterElementMutableData } from "./types";

export type FilterReducerActionType =
  | "clear"
  | "merge"
  | "reset"
  | "set-property";
export interface FilterReducerAction<T extends string> {
  type: FilterReducerActionType;
  payload: Partial<{
    name: T;
    update: Partial<IFilterElementMutableData>;
    new: IFilter<T>;
  }>;
}

function merge<T extends string>(
  prevState: IFilter<T>,
  newState: IFilter<T>
): IFilter<T> {
  return newState.map(newFilter => {
    const prevFilter = prevState.find(
      prevFilter => prevFilter.name === newFilter.name
    );
    if (!!prevFilter) {
      return {
        ...newFilter,
        active: prevFilter.active,
        value: prevFilter.value
      };
    }

    return newFilter;
  });
}

function setProperty<T extends string>(
  prevState: IFilter<T>,
  filter: T,
  updateData: Partial<IFilterElementMutableData>
): IFilter<T> {
  const field = prevState.find(f => f.name === filter);
  const updatedField = {
    ...field,
    ...updateData
  };

  return update(updatedField, prevState, (a, b) => a.name === b.name);
}

function reduceFilter<T extends string>(
  prevState: IFilter<T>,
  action: FilterReducerAction<T>
): IFilter<T> {
  switch (action.type) {
    case "set-property":
      return setProperty(prevState, action.payload.name, action.payload.update);
    case "merge":
      return merge(prevState, action.payload.new);
    case "reset":
      return action.payload.new;

    default:
      return prevState;
  }
}

export default reduceFilter;
