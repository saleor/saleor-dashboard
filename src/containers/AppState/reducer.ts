import IAppState, { AppError } from "./state";

export type AppStateReducerActionType = "displayError" | "displayLoader";

export interface AppStateReducerAction {
  payload: Partial<{
    error: AppError["type"];
    errorId: AppError["id"];
    value: boolean;
  }>;
  type: AppStateReducerActionType;
}

function displayError(
  prevState: IAppState,
  errorType: AppError["type"],
  errorId?: AppError["id"],
): IAppState {
  return {
    ...prevState,
    error: {
      id: errorId,
      type: errorType,
    },
    loading: false,
  };
}

function displayLoader(prevState: IAppState, value: boolean): IAppState {
  return {
    ...prevState,
    loading: value,
  };
}

function reduceAppState(
  prevState: IAppState,
  action: AppStateReducerAction,
): IAppState {
  switch (action.type) {
    case "displayError":
      return displayError(
        prevState,
        action.payload.error,
        action.payload.errorId,
      );
    case "displayLoader":
      return displayLoader(prevState, action.payload.value);
    default:
      return prevState;
  }
}

export default reduceAppState;
