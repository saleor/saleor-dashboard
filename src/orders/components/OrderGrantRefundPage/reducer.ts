import { OrderDetailsGrantRefundFragment } from "@dashboard/graphql";
import { exhaustiveCheck } from "@dashboard/utils/ts";

export interface ReducerOrderLine {
  selectedQuantity: number;
  availableQuantity: number;
  unitPrice: number;
}

export interface GrantRefundState {
  lines: Map<string, ReducerOrderLine>;
  refundShipping: boolean;
}

export type GrantRefundLineKeyValue = [string, ReducerOrderLine];

export type GrantRefundAction =
  | {
      type: "setQuantity";
      lineId: string;
      amount: number;
    }
  | {
      type: "setMaxQuantity";
      lineIds: string[];
    }
  | {
      type: "initState";
      state: GrantRefundState;
    }
  | {
      type: "toggleRefundShipping";
    };

export const getGrantRefundReducerInitialState = (
  order: OrderDetailsGrantRefundFragment,
): GrantRefundState => {
  const unfulfilledLines = order?.lines
    .filter(line => line.quantityToFulfill > 0)
    .map<GrantRefundLineKeyValue>(line => [
      line.id,
      {
        availableQuantity: line.quantity,
        unitPrice: line.unitPrice.gross.amount,
        selectedQuantity: 0,
      },
    ]);

  const fulfilmentLines = order.fulfillments
    .flatMap(fulfilment => fulfilment.lines)
    .map<GrantRefundLineKeyValue>(line => [
      line.id,
      {
        availableQuantity: line.quantity,
        unitPrice: line.orderLine.unitPrice.gross.amount,
        selectedQuantity: 0,
      },
    ]);

  return {
    lines: new Map([...unfulfilledLines, ...fulfilmentLines]),
    refundShipping: false,
  };
};

export const grantRefundDefaultState: GrantRefundState = {
  lines: new Map(),
  refundShipping: false,
};

export function grantRefundReducer(
  state: GrantRefundState,
  action: GrantRefundAction,
): GrantRefundState {
  switch (action.type) {
    case "setQuantity": {
      if (!state.lines.has(action.lineId)) {
        return state;
      }

      const line = state.lines.get(action.lineId);
      const newLines = new Map(state.lines);

      newLines.set(action.lineId, {
        ...line,
        selectedQuantity: action.amount,
      });

      return {
        ...state,
        lines: newLines,
      };
    }

    case "setMaxQuantity": {
      const newLines = new Map(state.lines);

      action.lineIds.forEach(lineId => {
        const line = state.lines.get(lineId);

        newLines.set(lineId, {
          ...line,
          selectedQuantity: line.availableQuantity,
        });
      });

      return {
        ...state,
        lines: newLines,
      };
    }

    case "initState": {
      return action.state;
    }

    case "toggleRefundShipping": {
      return {
        ...state,
        refundShipping: !state.refundShipping,
      };
    }

    default:
      exhaustiveCheck(action);
  }
}
