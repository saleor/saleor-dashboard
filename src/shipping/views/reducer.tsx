import {
  PostalCodeRuleInclusionTypeEnum,
  ShippingMethodWithPostalCodesFragment,
} from "@saleor/graphql";

export interface PostalCodesState {
  codesToDelete?: string[];
  havePostalCodesChanged?: boolean;
  inclusionType?: PostalCodeRuleInclusionTypeEnum;
  originalCodes?: ShippingMethodWithPostalCodesFragment["postalCodeRules"];
  postalCodeRules?: ShippingMethodWithPostalCodesFragment["postalCodeRules"];
}

function postalCodesReducer(
  prevState: PostalCodesState,
  newState: PostalCodesState,
) {
  return { ...prevState, ...newState };
}

export default postalCodesReducer;
