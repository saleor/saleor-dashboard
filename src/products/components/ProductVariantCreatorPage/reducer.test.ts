import {
  attributes,
  fourthStep,
  secondStep,
  thirdStep,
  warehouses
} from "./fixtures";
import reducer, {
  ProductVariantCreateReducerActionType,
  VariantField
} from "./reducer";

function execActions<TState, TAction>(
  initialState: TState,
  reducer: (state: TState, action: TAction) => TState,
  actions: TAction[]
): TState {
  return actions.reduce((acc, action) => reducer(acc, action), initialState);
}

describe("Reducer is able to", () => {
  it("select attribute values", () => {
    const state = execActions(secondStep, reducer, [
      {
        selectValue: {
          attributeId: attributes[0].id,
          valueId: attributes[0].values[0]
        },
        type: ProductVariantCreateReducerActionType.selectValue
      },
      {
        selectValue: {
          attributeId: attributes[0].id,
          valueId: attributes[0].values[6]
        },

        type: ProductVariantCreateReducerActionType.selectValue
      },
      {
        selectValue: {
          attributeId: attributes[1].id,
          valueId: attributes[1].values[1]
        },
        type: ProductVariantCreateReducerActionType.selectValue
      },
      {
        selectValue: {
          attributeId: attributes[1].id,
          valueId: attributes[1].values[3]
        },
        type: ProductVariantCreateReducerActionType.selectValue
      },
      {
        selectValue: {
          attributeId: attributes[3].id,
          valueId: attributes[3].values[0]
        },
        type: ProductVariantCreateReducerActionType.selectValue
      },
      {
        selectValue: {
          attributeId: attributes[3].id,
          valueId: attributes[3].values[4]
        },
        type: ProductVariantCreateReducerActionType.selectValue
      }
    ]);

    expect(state.attributes[0].values).toHaveLength(2);
    expect(state.attributes[1].values).toHaveLength(2);
    expect(state.attributes[2].values).toHaveLength(2);
    expect(state).toMatchSnapshot();
  });

  it("select price for all variants", () => {
    const price = "45.99";
    const state = execActions(thirdStep, reducer, [
      {
        applyPriceOrStockToAll: {
          mode: "all"
        },
        type: ProductVariantCreateReducerActionType.applyPriceToAll
      },
      {
        changeApplyPriceToAllValue: {
          price
        },
        type: ProductVariantCreateReducerActionType.changeApplyPriceToAllValue
      },
      {
        type: ProductVariantCreateReducerActionType.reload
      }
    ]);

    expect(state.price.mode).toBe("all");
    expect(state.price.value).toBe(price);
    expect(state).toMatchSnapshot();
  });

  it("select warehouses in which stock will be created", () => {
    const state = execActions(thirdStep, reducer, [
      {
        changeWarehouses: {
          warehouseId: warehouses[0].id
        },
        type: ProductVariantCreateReducerActionType.changeWarehouses
      },
      {
        changeWarehouses: {
          warehouseId: warehouses[2].id
        },
        type: ProductVariantCreateReducerActionType.changeWarehouses
      }
    ]);

    expect(state.warehouses).toHaveLength(2);
    expect(state).toMatchSnapshot();
  });

  it("select stock for all variants", () => {
    const quantity = 45;
    const warehouseIndex = 1;
    const state = execActions(thirdStep, reducer, [
      {
        applyPriceOrStockToAll: {
          mode: "all"
        },
        type: ProductVariantCreateReducerActionType.applyStockToAll
      },
      {
        changeApplyStockToAllValue: {
          quantity,
          warehouseIndex
        },
        type: ProductVariantCreateReducerActionType.changeApplyStockToAllValue
      },
      {
        type: ProductVariantCreateReducerActionType.reload
      }
    ]);

    expect(state.stock.mode).toBe("all");
    expect(state.stock.value[warehouseIndex]).toBe(quantity);
    expect(state).toMatchSnapshot();
  });

  it("select price to each attribute value", () => {
    const attribute = thirdStep.attributes[0];
    const value = 45.99;
    const state = execActions(thirdStep, reducer, [
      {
        applyPriceOrStockToAll: {
          mode: "attribute"
        },
        type: ProductVariantCreateReducerActionType.applyPriceToAll
      },
      {
        changeApplyPriceOrStockToAttributeId: {
          attributeId: attribute.id
        },
        type:
          ProductVariantCreateReducerActionType.changeApplyPriceToAttributeId
      },
      {
        changeAttributeValuePrice: {
          price: value.toString(),
          valueId: attribute.values[0]
        },
        type: ProductVariantCreateReducerActionType.changeAttributeValuePrice
      },
      {
        changeAttributeValuePrice: {
          price: (value + 6).toString(),
          valueId: attribute.values[1]
        },
        type: ProductVariantCreateReducerActionType.changeAttributeValuePrice
      },
      {
        type: ProductVariantCreateReducerActionType.reload
      }
    ]);

    expect(state.price.mode).toBe("attribute");
    expect(state.price.values).toHaveLength(
      state.attributes.find(attribute => state.price.attribute === attribute.id)
        .values.length
    );
    expect(state).toMatchSnapshot();
  });

  it("select stock to each attribute value", () => {
    const attribute = thirdStep.attributes[0];
    const quantity = 13;
    const state = execActions(thirdStep, reducer, [
      {
        applyPriceOrStockToAll: {
          mode: "attribute"
        },
        type: ProductVariantCreateReducerActionType.applyStockToAll
      },
      {
        changeApplyPriceOrStockToAttributeId: {
          attributeId: attribute.id
        },
        type:
          ProductVariantCreateReducerActionType.changeApplyStockToAttributeId
      },
      {
        changeAttributeValueStock: {
          quantity,
          valueId: attribute.values[0],
          warehouseIndex: 0
        },
        type: ProductVariantCreateReducerActionType.changeAttributeValueStock
      },
      {
        changeAttributeValueStock: {
          quantity: quantity + 6,
          valueId: attribute.values[1],
          warehouseIndex: 0
        },
        type: ProductVariantCreateReducerActionType.changeAttributeValueStock
      },
      {
        type: ProductVariantCreateReducerActionType.reload
      }
    ]);

    expect(state.stock.mode).toBe("attribute");
    expect(state.stock.values).toHaveLength(
      state.attributes.find(attribute => state.stock.attribute === attribute.id)
        .values.length
    );
    expect(state).toMatchSnapshot();
  });

  it("modify individual variant price", () => {
    const field: VariantField = "price";
    const value = "49.99";
    const variantIndex = 3;

    const state = execActions(fourthStep, reducer, [
      {
        changeVariantData: {
          field,
          value,
          variantIndex
        },
        type: ProductVariantCreateReducerActionType.changeVariantData
      }
    ]);

    expect(state.variants[variantIndex].priceOverride).toBe(value);
    expect(state.variants[variantIndex - 1].priceOverride).toBe(
      fourthStep.variants[variantIndex - 1].priceOverride
    );
    expect(state).toMatchSnapshot();
  });

  it("modify individual variant stock", () => {
    const quantity = 5;
    const variantIndex = 3;

    const state = execActions(fourthStep, reducer, [
      {
        changeVariantStockData: {
          stock: {
            quantity,
            warehouse: warehouses[0].id
          },
          variantIndex
        },
        type: ProductVariantCreateReducerActionType.changeVariantStockData
      }
    ]);

    expect(state.variants[variantIndex].stocks[0].quantity).toBe(quantity);
    expect(state.variants[variantIndex - 1].stocks[0].quantity).toBe(
      fourthStep.variants[variantIndex - 1].stocks[0].quantity
    );
    expect(state).toMatchSnapshot();
  });

  it("delete variant", () => {
    const variantIndex = 3;

    const state = execActions(fourthStep, reducer, [
      {
        deleteVariant: {
          variantIndex
        },
        type: ProductVariantCreateReducerActionType.deleteVariant
      }
    ]);

    expect(state.variants.length).toBe(fourthStep.variants.length - 1);
  });
});
