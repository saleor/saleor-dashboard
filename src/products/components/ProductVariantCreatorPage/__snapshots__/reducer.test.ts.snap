// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer is able to modify individual variant stock 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "channelListings": Array [],
  "price": Object {
    "attribute": "attr-2",
    "mode": "attribute",
    "value": "",
    "values": Array [
      Object {
        "slug": "val-2-2",
        "value": "24.99",
      },
      Object {
        "slug": "val-2-4",
        "value": "26.99",
      },
    ],
  },
  "stock": Object {
    "attribute": "attr-4",
    "mode": "attribute",
    "value": Array [],
    "values": Array [
      Object {
        "slug": "val-4-1",
        "value": Array [
          50,
          20,
          45,
          75,
        ],
      },
      Object {
        "slug": "val-4-5",
        "value": Array [
          80,
          50,
          85,
          105,
        ],
      },
    ],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 50,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 20,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 75,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 80,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 50,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 85,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 105,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 50,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 20,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 75,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 5,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 50,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 85,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 105,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 50,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 20,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 75,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 80,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 50,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 85,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 105,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 50,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 20,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 75,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 80,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 50,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 85,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 105,
          "warehouse": "wh-4",
        },
      ],
    },
  ],
  "warehouses": Array [
    "wh-1",
    "wh-2",
    "wh-3",
    "wh-4",
  ],
}
`;

exports[`Reducer is able to select attribute values 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "channelListings": Array [],
  "price": Object {
    "attribute": undefined,
    "mode": "all",
    "value": "",
    "values": Array [],
  },
  "stock": Object {
    "attribute": undefined,
    "mode": "all",
    "value": Array [],
    "values": Array [],
  },
  "variants": Array [],
  "warehouses": Array [],
}
`;

exports[`Reducer is able to select stock for all variants 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "channelListings": Array [],
  "price": Object {
    "attribute": undefined,
    "mode": "all",
    "value": "",
    "values": Array [],
  },
  "stock": Object {
    "attribute": undefined,
    "mode": "all",
    "value": Array [
      0,
      45,
      0,
      0,
    ],
    "values": Array [],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 0,
          "warehouse": "wh-1",
        },
        Object {
          "quantity": 45,
          "warehouse": "wh-2",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-3",
        },
        Object {
          "quantity": 0,
          "warehouse": "wh-4",
        },
      ],
    },
  ],
  "warehouses": Array [
    "wh-1",
    "wh-2",
    "wh-3",
    "wh-4",
  ],
}
`;

exports[`Reducer is able to select stock to each attribute value 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "channelListings": Array [],
  "price": Object {
    "attribute": undefined,
    "mode": "all",
    "value": "",
    "values": Array [],
  },
  "stock": Object {
    "attribute": "attr-1",
    "mode": "attribute",
    "value": Array [
      0,
      0,
      0,
      0,
    ],
    "values": Array [
      Object {
        "slug": "val-1-1",
        "value": Array [
          13,
        ],
      },
      Object {
        "slug": "val-1-7",
        "value": Array [
          19,
        ],
      },
    ],
  },
  "variants": Array [
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 13,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 13,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 13,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-1",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 13,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 19,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-2",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 19,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-1",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 19,
          "warehouse": "wh-1",
        },
      ],
    },
    Object {
      "attributes": Array [
        Object {
          "id": "attr-1",
          "values": Array [
            "val-1-7",
          ],
        },
        Object {
          "id": "attr-2",
          "values": Array [
            "val-2-4",
          ],
        },
        Object {
          "id": "attr-4",
          "values": Array [
            "val-4-5",
          ],
        },
      ],
      "sku": "",
      "stocks": Array [
        Object {
          "quantity": 19,
          "warehouse": "wh-1",
        },
      ],
    },
  ],
  "warehouses": Array [
    "wh-1",
    "wh-2",
    "wh-3",
    "wh-4",
  ],
}
`;

exports[`Reducer is able to select warehouses in which stock will be created 1`] = `
Object {
  "attributes": Array [
    Object {
      "id": "attr-1",
      "values": Array [
        "val-1-1",
        "val-1-7",
      ],
    },
    Object {
      "id": "attr-2",
      "values": Array [
        "val-2-2",
        "val-2-4",
      ],
    },
    Object {
      "id": "attr-4",
      "values": Array [
        "val-4-1",
        "val-4-5",
      ],
    },
  ],
  "channelListings": Array [],
  "price": Object {
    "attribute": undefined,
    "mode": "all",
    "value": "",
    "values": Array [],
  },
  "stock": Object {
    "attribute": undefined,
    "mode": "all",
    "value": Array [
      0,
      0,
    ],
    "values": Array [],
  },
  "variants": Array [],
  "warehouses": Array [
    "wh-2",
    "wh-4",
  ],
}
`;
